torch/distributions/constraint_registry.py

```diff
@@ -0,0 +1,203 @@
+   implement ``.log_abs_det_jacobian()``.
+
+The ``transform_to()`` registry is useful for performing unconstrained
+optimization on constrained parameters of probability distributions, which are
+indicated by each distribution's ``.params`` dict. These transforms often
+overparameterize a space in order to avoid rotation; they are thus more
+suitable for coordinate-wise optimization algorithms like Adam::
+
+    loc = Variable(torch.zeros(100), requires_grad=True)
+    unconstrained = Variable(torch.zeros(100), requires_grad=True)
+    scale = transform_to(Normal.params['scale'])(unconstrained)
+    loss = -Normal(loc, scale).log_prob(data).sum()
+
+The ``biject_to()`` registry is useful for Hamiltonian Monte Carlo, where
+samples from a probability distribution with constrained ``.support`` are
+propagated in an unconstrained space, and algorithms are typically rotation
+invariant.::
+
+    dist = Exponential(rate)
+    unconstrained = Variable(torch.zeros(100), requires_grad=True)
+    sample = biject_to(dist.support)(unconstrained)
+    potential_energy = -dist.log_prob(sample).sum()
+
+.. note::
+
+    An example where ``transform_to`` and ``biject_to`` differ is
+    ``constraints.simplex``: ``transform_to(constraints.simplex)`` returns a
+    :class:`~torch.distributions.transforms.BoltzmannTransform` that simply
+    exponentiates and normalizes its inputs; this is a cheap and mostly
+    coordinate-wise operation appropriate for algorithms like SVI. In
+    contrast, ``biject_to(constraints.simplex)`` returns a
+    :class:`~torch.distributions.transforms.StickBreakingTransform` that
+    bijects its input down to a one-fewer-dimensional space; this a more
+    expensive less numerically stable transform but is needed for algorithms
+    like HMC.
+
+The ``biject_to`` and ``transform_to`` objects can be extended by user-defined
+constraints and transforms using their ``.register()`` method either as a
+function on singleton constraints::
+
+    transform_to.register(my_constraint, my_transform)
+
+or as a decorator on parameterized constraints::
+
+    @transform_to.register(MyConstraintClass)
+    def my_factory(constraint):
+        assert isinstance(constraint, MyConstraintClass)
+        return MyTransform(constraint.param1, constraint.param2)
+
+You can create your own registry by creating a new :class:`ConstraintRegistry`
+object.
+"""
+
+from torch.distributions import constraints, transforms
+
+__all__ = [
+    'ConstraintRegistry',
+    'biject_to',
+    'transform_to',
+]
+
+
+class ConstraintRegistry(object):
+    """
+    Registry to link constraints to transforms.
+    """
+    def __init__(self):
+        self._registry = {}
+
+    def register(self, constraint, transform=None):
+        """
+        Registers a :class:`~torch.distributions.constraints.Constraint`
+        subclass or singleton object in this registry. Usage as decorator::
+
+            @my_registry.register(MyConstraintClass)
```

Comment:
Is it expected that people will be creating more registries?

Reply:
I don't know. I needed two, and @martinjankowiak noted that it is useful to tweak which transforms are used for each constraint. One plausible application is to implement a registry `transform_from_unit_interval_to()` whereas the builtin registries both transform from `constraints.real` to various other constraints.

But I'm ok making this private if you'd prefer a narrower interface.